torchrl/modules/models/multiagent.py

```diff
@@ -81,48 +88,77 @@ def exec_module(params, *input):
 
         return torch.vmap(exec_module, *args, **kwargs)
 
-    def forward(self, *inputs: Tuple[torch.Tensor]) -> torch.Tensor:
-        if len(inputs) > 1:
-            inputs = torch.cat([*inputs], -1)
+    def _in_dim(self, input: Tuple[...]):
+        # We must define a structure that resembles the input data with the
+        # agent dims in it, smth like (None,) if there is a single input
+        # of (None, (None, None)) for an LSTM
+        in_dim = self.__dict__.get("_in_dim_value", None)
+        if in_dim is None:
+            in_dim = tree_map(
+                lambda x: None if self.centralised else self.agent_dim, input
+            )
+            self.__dict__["_in_dim_value"] = in_dim
+        return in_dim
+
+    def _expand_centralized_output(self, output):
+        # If the parameters are shared, and it is centralised, all agents will have the same output
+        # We expand it to maintain the agent dimension, but values will be the same for all agents
+        def expand_output(output):
+            n_agent_outputs = output.shape[-1]
+            output = output.view(*output.shape[:-1], n_agent_outputs)
+            output = output.unsqueeze(-2)
+            output = output.expand(*output.shape[:-2], self.n_agents, n_agent_outputs)
+            return output
+
+        if self.num_outputs == 1:
+            return expand_output(output)
         else:
-            inputs = inputs[0]
+            return tree_map(expand_output, output)
 
+    def forward(self, *inputs: Tuple[torch.Tensor]) -> torch.Tensor:
         inputs = self._pre_forward_check(inputs)
         # If parameters are not shared, each agent has its own network
         if not self.share_params:
             if self.centralised:
+                tree_map(lambda x: print(x.shape) if x is not None else None, inputs)
                 output = self.vmap_func_module(
-                    self._empty_net, (0, None), (-2,), randomness=self._vmap_randomness
-                )(self.params, inputs)
+                    self._empty_net,
+                    (0, *self._in_dim(inputs)),
+                    (-2,) * self.num_outputs,
+                    randomness=self._vmap_randomness,
+                )(self.params, *inputs)
             else:
                 output = self.vmap_func_module(
                     self._empty_net,
-                    (0, self.agent_dim),
-                    (-2,),
+                    (0, *self._in_dim(inputs)),
+                    (-2,) * self.num_outputs,
                     randomness=self._vmap_randomness,
-                )(self.params, inputs)
+                )(self.params, *inputs)
 
         # If parameters are shared, agents use the same network
+        elif not self.centralised:
+            output = self.vmap_func_module(
+                self._empty_net,
+                (None, *self._in_dim(inputs)),
+                (-2,) * self.num_outputs,
+                randomness=self._vmap_randomness,
```

Comment:
I have a question about the randomness

```python
    @property
    def _vmap_randomness(self):
        if self.initialized:
            return "error"
        return "same"
```

Why would it be "error" when the net is initialized? and "same" otherwise? this code needs comments